share/ipld/get.go

```diff
@@ -216,17 +225,44 @@ func GetLeavesByNamespace(
 	// on the level above, the length of the Row is passed in as maxShares
 	leaves := make([]ipld.Node, maxShares)
 
+	var leftProofs, rightProofs [][]byte
+	if collectProofs {
+		// TODO: (@walldiss) this is massively overallocating and should be optimized later with clever append
+		leftProofs = make([][]byte, BatchSize((maxShares)))
+		rightProofs = make([][]byte, BatchSize((maxShares)))
+	}
+
 	for {
 		select {
 		case j, ok := <-jobs:
 			if !ok {
 				// if there were no leaves under the given root in the given namespace,
 				// both return values are nil. otherwise, the error will also be non-nil.
 				if bounds.lowest == int64(maxShares) {
-					return nil, retrievalErr
+					return &LeavesWithProofs{}, retrievalErr
+				}
+
+				var proofs [][]byte
+				// left side traversed in bottom-up order
+				for i := range leftProofs {
+					if leftProofs[i] != nil {
+						proofs = append(proofs, leftProofs[i])
+					}
+				}
+				// right side of the tree will be traversed from up to bottom,
+				// so append in reversed order
+				for i := range rightProofs {
+					if rightProofs[len(rightProofs)-i-1] != nil {
+						proofs = append(proofs, rightProofs[len(rightProofs)-i-1])
+					}
```

Comment:
Can you check if this loop would be more readable if you use a normal for loop that just counts down instead of up? but if you don't like it we can keep it this way

Reply:
Looks cleaner with counting down, thank you.

Reply:
np